bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
Hi, I'm definitely happy to change the concepts/approach, but I think the end result is really helpful - I find it super weird that our default charts let you pan a bar chart, heat map beyond their max and min. On continuous axes you might be outside the data, but at least it has meaning on a categorical axis it has no meaning. But, for example, on a heat map you do still want panning and zooming, just not beyond the extents. 

I was tempted to implement, that if you don't set one, it gets set to `bound_lower=factors[0]` and `bound_upper=factors[-1]`

Just to re-state, I'm not disagreeing that `upper` and `lower` are conceptually odd, but I'd love to see us enable something like this, and perhaps push it further than this PR does.